server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+                        throw new ParseException(
+                            "Sets can only be filtered using sets of the same type!");
+                      }
+                    }
+
+                    modifiers.add(new OperatorDto(UcdlToken.FILTER, filters));
+                    modifierList = modifierList.jjtGetChild(2);
+                    parameters.replace("this", thisType);
+                  }
+                }
+
+                return new SetDto(UcdlToken.RESOURCE_SET, setType, setName, modifiers);
+
+              case "ELEMENTINSET":
+                ElementDto element = (ElementDto) buildAst(root.jjtGetChild(0), parameters);
+                SetDto set = (SetDto) buildAst(root.jjtGetChild(1).jjtGetChild(0), parameters);
+
+                if (element.type() != set.type()) {
+                  throw new ParseException("Elements can only be in Sets of the same type!");
+                }
+
+                List<AbstractSyntaxTreeDto> params = new ArrayList<>();
+                params.add(element);
+                params.add(set);
+
+                AbstractSyntaxTreeDto elementInSet = new OperatorDto(UcdlToken.IN, params);
+
+                if (root.jjtGetNumChildren() == 2) {
+                  return elementInSet;
+                }
+                params = new ArrayList<>();
+                params.add(elementInSet);
+                params.add(buildAst(root.jjtGetChild(2).jjtGetChild(0), parameters));
+                return switch (((SimpleNode) root.jjtGetChild(2)).jjtGetFirstToken().image) {
+                  case "implies", "->" -> new OperatorDto(UcdlToken.IMPLIES, params);
+                  case "or", "||" -> new OperatorDto(UcdlToken.OR, params);
+                  case "and", "&&" -> new OperatorDto(UcdlToken.AND, params);
+                  default -> throw new IllegalStateException();
+                };
+              case "EQUATION":
+                params = new ArrayList<>();
+                Node firstElement = root.jjtGetChild(0);
+                Node secondElement = root.jjtGetChild(1).jjtGetChild(1);
+                String comparator =
+                    ((SimpleNode) root.jjtGetChild(1).jjtGetChild(0)).jjtGetFirstToken().image;
+
+                params.add(buildAst(firstElement, parameters));
+                params.add(buildAst(secondElement, parameters));
+                UcdlToken token = getComparatorToken(comparator, params);
+                return new OperatorDto(token, params);
+              default:
+                throw new IllegalStateException();
+            }
+          default:
+            throw new IllegalStateException();
+
+        }
+      }
+      //includes cases: "CONTROLSEQUENCELIST", "CONTROLSEQUENCERETURN",
+      // "OPTIONALIMPLIES", "OPTIONALOR", "OPTIONALAND",
+      // "ELEMENTINSET", "EQUATION", "ELEMENTEQUATION", "ELEMENTATTRIBUTELIST",
+      // "SETNAME", "SETMODIFICATION", "NUMBERLIST", "ATTRIBUTE", "FILTERLIST"
```

Comment:
You don't mean the default case includes those cases, do you?

Reply:
It does.
Those are part of other cases and do not align with the general AST structure and as the method is required to return a valid AST it would not make sense to allow for those cases to be executed directly.
However they are still listed to make sure that every possible case is taken care of.

Reply:
Ah, so basically those are valid tokens, but we shouldn't run into them here. And if we do, something went completely wrong. Thanks for the explanation!